lms/static/scripts/frontend_apps/components/dashboard/DashboardActivityFilters.tsx

```diff
@@ -23,6 +28,62 @@ export type DashboardActivityFiltersProps = {
   onClearSelection?: () => void;
 };
 
+function usePaginatedAPIFetch<
+  Prop extends string,
+  T extends any[],
+  K extends { [P in Prop]: T } & { pagination: Pagination },
+>(
+  prop: Prop,
+  path: string,
+  filters: Record<string, string | string[] | undefined>,
+): [T | undefined, FetchResult<K>, (url: string) => void] {
+  const prevFiltersRef = useRef<typeof filters>(filters);
+  const [urlFromPagination, setURLFromPagination] = useState<string>();
+  const [currentList, setCurrentList] = useState<T>();
+
+  const result = useAPIFetch<K>(
+    urlFromPagination ?? path,
+    !urlFromPagination ? filters : undefined,
+  );
+
+  // Every time data is loaded, append the result to the list
+  // TODO This assumes every page is loaded only once.
+  useEffect(() => {
+    // @ts-expect-error todo investigate this type error
+    setCurrentList(prev =>
+      result.data ? [...(prev ?? []), ...result.data[prop]] : prev,
+    );
+  }, [prop, result.data]);
+
+  // Every time filters change, discard previous pages and start from scratch
+  useEffect(() => {
+    const prevFilters = JSON.stringify(prevFiltersRef.current);
+    const newFilters = JSON.stringify(filters);
+
+    if (prevFilters !== newFilters) {
```

Comment:
I'm not sure if this is the most efficient way to deep-compare if the filters changed.

Considering all are memozied in the component, perhaps a reference check would do (`prevFilters === newFilters`), but that implies this hook is making assumptions on how it is used, making it harder to eventually extract.